File: comp-sci/bst/python/bst.py
import os

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root:Node = None

    def insert(self, data):
        if self.root is None:
            self.root = Node(data)
        else:
            self._insert(self.root, data)
    def _insert(self, cur:Node, data):
        if data < cur.data:
            if cur.left is None:
                cur.left = Node(data)
            else:
                self._insert(cur.left, data)
        elif data > cur.data:
            if cur.right is None:
                cur.right = Node(data)
            else:
                self._insert(cur.right, data)
        else:
            raise KeyError("Duplicate data exists in BST.")

    def exists(self, data) -> bool:
        return self._exists(self.root, data)
    def _exists(self, cur, data) -> bool:
        if cur is None:
            return False
        if cur.data == data:
            return True
        elif data < cur.data:
            return self._exists(cur.left, data)
        else:
            return self._exists(cur.right, data)

    def delete(self, data):
        pass

    def minimum(self):
        cur = self.root
        while cur.left is not None:
            cur = cur.left
        return cur.data

    def maximum(self):
        cur = self.root
        while cur.right is not None:
            cur = cur.right
        return cur.data

    def inorder_traverse(self) -> list:
        result = []    # ref will be passed around to collect
        self._inorder(self.root, result)
        return result
    def _inorder(self, cur:Node, result):
        if cur is not None:
            self._inorder(cur.left, result)
            result.append(cur.data)
            self._inorder(cur.right, result)

    def preorder_traverse(self) -> list:
        result = []
        self._preorder(self.root, result)
        return result
    def _preorder(self, cur:Node, result):
        if cur is not None:
            result.append(cur.data)
            self._preorder(cur.left, result)
            self._preorder(cur.right, result)

    def postorder_traverse(self) -> list:
        result = []
        self._postorder(self.root, result)
        return result
    def _postorder(self, cur:Node, result):
        if cur is not None:
            self._postorder(cur.left, result)
            self._postorder(cur.right, result)
            result.append(cur.data)

    def height(self) -> int:
        return self._height(self.root)
    def _height(self, node) -> int:
        if node is None:
            return -1
        left_height = self._height(node.left)
        right_height = self._height(node.right)
        return 1 + max(left_height, right_height)

    def size(self):
        return len(self.inorder_traverse())
    
        

File: comp-sci/bst/cpp-with-delete/bst.cpp
#include "bst.hpp"
#include <algorithm>
#include <cstdlib>
#include <cstdio>    // for perror
#include <cerrno>    // for global errno
#include <iostream>

BST::BST() : root(nullptr) {
}


// returns:
//   - true: successfully added.
//   - false: failed.  error occured.  (duplicate key)
//
bool
BST::insert(int data) {
    if (root == nullptr) {
        root = std::make_unique<Node>(data);
        return true;
    }

    Node *cur = root.get();
    Node *par = nullptr;

    while (cur != nullptr) {
        par = cur;
        if (data < cur->data) {
            cur = cur->left.get();
        } else if (data > cur->data ) {
            cur = cur->right.get();
        } else {
            // DUPLICATE!
            return false;
        }
    }

    if (data < par->data) {
        par->left = std::make_unique<Node>(data);
    } else {
        par->right = std::make_unique<Node>(data);
    }
    return true;
}

bool
BST::exists(int data) const {
    if (root == nullptr)
        return false;

    Node *cur = root.get();
    while ( cur != nullptr ) {
        if (data < cur->data) {
            cur = cur->left.get();
        } else if ( data > cur->data) {
            cur = cur->right.get();
        } else {
            return true;
        }
    }
    return false;
}

// returns:
//   - true: successfully deleted.
//   - false: failed. node not found.
//
bool
BST::remove(int data) {
    if (root == nullptr)
        return false;

    // Find the node to delete and its parent
    //   - Doing this will simplify tracing parent.
    //   - ownership switch (ptr->reset() ) will handle memory dealloc and parent pair.
    //
    // Summary:
    //    - *node_ptr : will adjust ownership.
    //    - *node     : will deference the pointer.
    //
    std::unique_ptr<Node> *node_ptr = &root;

    while (*node_ptr != nullptr) {
        Node *node = node_ptr->get();

        if (data < node->data) {
            node_ptr = &(node->left);
        } else if (data > node->data) {
            node_ptr = &(node->right);
        } else {
            // Found the node to delete

            // Case 1: Node has no children (leaf node)
            if (node->left == nullptr && node->right == nullptr) {
                node_ptr->reset();
                return true;
            }

            // Case 2: Node has only right child
            if (node->left == nullptr) {
                *node_ptr = std::move(node->right);
                return true;
            }

            // Case 3: Node has only left child
            if (node->right == nullptr) {
                *node_ptr = std::move(node->left);
                return true;
            }

            // Case 4: Node has two children
            // Find the inorder successor (minimum in right subtree)
            Node *successor_parent = node;
            Node *successor = node->right.get();
            while (successor->left != nullptr) {
                successor_parent = successor;
                successor = successor->left.get();
            }

            // Copy successor's data to current node
            node->data = successor->data;

            // Delete the successor (which has at most one child - right child)
            if (successor_parent == node) {
                // Special case:  Successor is the right child of node
                //
                //    node
                //       \
                //       successor
                //         \
                //          TREE
                //
                node->right = std::move(successor->right);
            } else {
                // Successor is in the left subtree of node's right child
                //successor_parent->left = std::move(successor->left);    // intentional fail, for unittest
                successor_parent->left = std::move(successor->right);
            }

            return true;
        }
    }

    // Node not found
    return false;
}

// PRE: I don't handle empty-tree's minimum.
//   - it will crash.
int
BST::minimum() const {
    if (root == nullptr) {
        perror("ERROR: min() for empty tree.");
        exit(1);
    }
    Node *cur = root.get();
    while (cur->left != nullptr)
        cur = cur->left.get();
    return cur->data;
}
// PRE: I don't handle empty-tree's maximum.
//   - it will crash.
int
BST::maximum() const {
    if (root == nullptr) {
        perror("ERROR: min() for empty tree.");
        exit(1);
    }
    Node *cur = root.get();
    while (cur->right != nullptr)
        cur = cur->right.get();
    return cur->data;
}

int
BST::height() const {
    return _height(root.get());
}

int
BST::_height(Node *subtree_root) const {
    Node *node = subtree_root;
    if (node == nullptr)
        return -1;

    int left_height = _height(node->left.get());
    int right_height  = _height(node->right.get());
    return 1 + std::max(left_height, right_height);
}

void
BST::printTree() const {
    _printTree(root.get(), 0);
}

void
BST::_printTree(Node* node, int level) const {
    if (node == nullptr) {
        return;
    }

    _printTree(node->right.get(), level + 1);
    for (int i = 0; i < level; i++) {
        std::cout << "    ";
    }
    std::cout << node->data << std::endl;
    _printTree(node->left.get(), level + 1);
}
